### DISCUSSION TITLE
Module federation with "path imports" e.g. @my-package/x/y

### DISCUSSION THREAD

**Discussion by lee-borsboom** [webpack/webpack#13612] [Q&A]
Hey everyone,

I'm doing a proof of concept application of module federation before adopting it for usage on a significant existing solution that currently uses iframes to achieve a "micro app architecture", but I'm getting stuck on one point. At a high level, I have the following:

app1 (remote)
app2 (remote)
sharedLibrary

Throughout app1 and app2 we import from the shared library like:

`import X from @my-scope/shared-library/myComponent`

I have both in-house libraries that we use like this and third-party dependencies (e.g. Material UI)

Unless I have missed something, it seems that by doing it this way, we'll have to configure each component as a shared module in the Module Federation plugin which represents a maintenance burden, made yet worse because I want them to be singletons so the config is even more verbose. e.g.

```
      new ModuleFederationPlugin({
        exposes: { './App': './src/app.js' },
        filename: 'remoteEntry.js',
        name: 'mfApp2',
        shared: [{
          react: {
            singleton: true,
            requiredVersion: dependencies.react
          },
          'react-router-dom': {
            singleton: true,
            requiredVersion: dependencies['react-router-dom']
          },
          '@my-scope/shared-library/card': {
            singleton: true,
            requiredVersion: dependencies['@my-scope/shared-library']
          },
          '@my-scope/shared-library/button': {
            singleton: true,
            requiredVersion: dependencies['@my-scope/shared-library']
          },
          // ... many, many more components
        }]
    })
```

Is there an alternative configuration option which would allow me to do something like:

```
'@my-scope/shared-library/*': {
   singleton: true,
   requiredVersion: dependencies['@my-scope/shared-library']
}
```

I believe webpack externals supports the use of regular expressions for the module names which facilitates things like this. I'm also wondering whether perhaps a viable solution is to in fact use webpack externals for the shared libraries and module federation for the apps but that clearly has its drawbacks and at best undermines the utility of module federation.

I specifically want to avoid having to have an index.js in shared-library that all the apps use via a destructuring import e.g.

`import { X } from '@my-company/shared-library'`

When `shared-library` is large, imports like in the above significantly degrade the developer experience, resulting in slow start up times because the entirety of `shared-library` must be loaded up front.

**Comment by lee-borsboom** [comment]
Also, as mentioned above, we also use path imports from a third party library (Material UI). Given the following scenario:

@my-company/shared-library/x <--- imports @my-company/shared-library/y
@my-company/shared-library/y <--- imports @material-ui/core/Card

app1 <--- imports @my-company/shared-library/x
app2 <--- imports @my-company/shared-library/y

@material-ui/core/Card is then loaded once for app1 and again for app2, unless app1 and app2 both explicitly list @material-ui/core/Card as `shared` in their webpack config. This happens even if both app1 and app2 have their webpack config explicitly listing _both_ x and y, even though they're only using one or the other directly. Thus it seems we then need to not only define every single component from sharedLibrary as `shared` for every app, we also need to define every single transitive dependency.

I'd love to be proven wrong, hopefully I've just messed up some config somewhere.

**Comment by vgarba1** [comment]
Had the same problem when I initially set up. You were super close! You can set up shared configs with keys ending in '/'. Inside the MFE plugin. ending with '/' with be used to match as a prefix on requests ( with some exceptions ). so `@myCommonLib/: { ... config }` will be matched with `@myCommonLib/someFile.js` and exposed to sharing.

**Comment by vgarba1** [comment]
References -
ProvideSharedPlugin: https://github.com/webpack/webpack/blob/400a0f94ab45ca20b10f219c8311e87d3d3f108c/lib/sharing/ProvideSharedPlugin.js#L170
ConsumeSharedPlugin: https://github.com/webpack/webpack/blob/400a0f94ab45ca20b10f219c8311e87d3d3f108c/lib/sharing/ConsumeSharedPlugin.js#L269
